### PR TITLE
feat: Refresh app tokens

### PR DESCRIPTION
**Pull request by sicarul**
When running long initial syncs using Github app private keys, the tokens expire in 1 hour, which makes the process fail after this timespan. This PR separates the app tokens into a new function `refresh_app_token` that can get called again in `get_next_auth_token`, i've tested it for my use case and it no longer fails after 1 hour.

### CODE REVIEW

**Comment by ericboucher** [comment]
Thanks @sicarul for your PR. I'll have a look and test it. In the meantime, could you make sure mypy and other tests are passing?

**Comment by sicarul** [comment]
I've fixed the mypy error, it seems it doesn't like me defining the type of the variable on __init__ for some reason

File: tap_github/authenticator.py [hunk 1]
```diff
@@ -211,7 +211,7 @@ def __init__(self, stream: RESTStream) -> None:
         self.active_token: Optional[TokenRateLimit] = (
             choice(list(self.tokens_map.values())) if len(self.tokens_map) else None
         )
-        self.last_private_key_token_refresh: datetime = None
+        self.last_private_key_token_refresh = None
```
[the remark this answers is not in the document]
Comment (sicarul) [reply tap_github/authenticator.py:214]: If i do that i get this error:

> tap_github/authenticator.py:214: error: Attribute "last_private_key_token_refresh" already defined on line 146  [no-redef]

But i haven't defined it anywhere else, i do set it on refresh_app_token, any ideas how i can solve this to add typing?

File: tap_github/authenticator.py [hunk 2]
```diff
@@ -152,7 +142,32 @@ def prepare_tokens(self) -> Dict[str, TokenRateLimit]:
                 app_token = generate_app_access_token(
                     github_app_id, github_private_key, github_installation_id or None
                 )
-                available_tokens = available_tokens + [app_token]
+                # Get rate_limit_buffer
+                rate_limit_buffer = self._config.get("rate_limit_buffer", None)
+                # Update last refresh timestamp
+                self.last_private_key_token_refresh = datetime.now()
+
+                return TokenRateLimit(app_token, rate_limit_buffer)
```
Comment (ericboucher) [inline tap_github/authenticator.py:150]: This is a change in behavior as it returns only this token now.

**Review by ericboucher — COMMENTED** [review]
Thanks @sicarul for opening this PR and sorry it took so long to get back to you.

The logic looks great overall. However, I think it would make more sense to add the new timestamp check within `TokenRateLimit` or something along these lines. This way, you will be able to keep running the tap with multiple token types if needed and leverage the existing calls to "is_valid" and other rate-limiting checks.

Let me know what you think!

**Comment by sicarul** [comment]
Hey @ericboucher i don't understand how would the timestamp check work in that class, it does not handle any kind of token refresh and it's used for all types of tokens, by putting the logic there it seems i would be generating a spaghetti of code to me, but maybe i'm just not understanding how do you want me to change this.

**Comment by laurentS** [comment]
If I may chime in, I did find it a bit difficult to follow the overall logic (some high level comments would help a lot). Am I getting this right?
- the tap first tries to load `GITHUB_TOKEN...` tokens passed as env vars, then loads a private key based one, only if there are no other tokens (during init). So you can't cumulate API tokens+private key anymore (`self.token_map` never gets the app token now)?
- then as it uses tokens, we call `get_next_auth_token` if we run out of quota (which we know from getting an error from the server), to rotate the token (but this excludes the private key based one)
- then when getting `auth_headers` we would rotate the private key based token if that's what we're using.

If the above is correct:
- I find it confusing that there is a non-obvious side effect in `auth_headers` (of refreshing the token)
- I think it's a regression if we can't use both types of tokens together anymore (we might want to cumulate limits from multiple accounts).

**Comment by sicarul** [comment]
Got it, i'm going on a vacation tomorrow until July but when i come back i can check this feedback and apply it to get a better version out if it's fine with you all!

@laurents sorry if it's not clear, but this version will still use both types of tokens as i implemented it.

**Comment by sicarul** [comment]
Closing since this is being fixed in the new PR: https://github.com/MeltanoLabs/tap-github/pull/289

**Pull request closed without merging**